# The PayScale Dataset 2021

In [1]:
# Preliminary Data Exploration

In [2]:
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
raw_csv_path = './raw_csv/raw_college_salary_report_2021.csv'
raw_df = pd.read_csv(raw_csv_path)

In [3]:
raw_df.head(5)

Rank                                             Major Degree Type  \
0    -                                                 -           -   
1    1                             Petroleum Engineering   Bachelors   
2    2      Operations Research & Industrial Engineering   Bachelors   
3    3  Electrical Engineering & Computer Science (EECS)   Bachelors   
4    4                                Interaction Design   Bachelors   

                                    Early Career Pay  \
0  Median salary for alumni with 0-5 years experi...   
1                                            $93,200   
2                                            $84,800   
3                                           $108,500   
4                                            $68,300   

                                      Mid-Career Pay  \
0  Median salary for alumni with 10+ years experi...   
1                                           $187,300   
2                                           $170,400   
3                                           $159,300   
4                                           $155,800   

                                      % High Meaning  
0  % alumni who say their work makes the world a ...  
1                                                67%  
2                                                28%  
3                                                46%  
4                                                55%

In [4]:
raw_df.shape

(828, 6)

In [5]:
raw_df.describe()

Rank Major Degree Type Early Career Pay Mid-Career Pay % High Meaning
count   828   828         828              828            828            828
unique  507   828           2              329            507             73
top     397     -   Bachelors          $47,700        $85,900              -
freq      7     1         827               12              7             64

In [6]:
raw_df['Degree Type'].value_counts()  
# --> Only 'Bachelors' in 'Degree Type', except '-' in info-row.

Bachelors    827
-              1
Name: Degree Type, dtype: int64

**Check for duplicates**

In [7]:
# Check for duplicates:

print(f'{raw_df.duplicated().any() = }') 
# Returns true if any duplicates of a whole row

print(f'{raw_df["Major"].duplicated().any() = }')
# Returns true if any duplicates in column 'Major'

raw_df.duplicated().any() = False
raw_df["Major"].duplicated().any() = False


**Check index - reset if necessary**

In [8]:
raw_df.index.values  # --> Repeating index-pattern.

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24, 25,  0,  1,  2,  3,  4,  5,  6,  7,
        8,  9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24,
        0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24,  0,  1,  2,  3,  4,  5,  6,  7,  8,
        9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24,  0,
        1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24,  0,  1,  2,  3,  4,  5,  6,  7,  8,  9,
       10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24,  0,  1,
        2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17, 18,
       19, 20, 21, 22, 23, 24,  0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10,
       11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24,  0,  1,  2,
        3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19,
       20, 21, 22, 23, 24

In [9]:
raw_df.index.duplicated()

array([False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,

In [10]:
df = raw_df.reset_index(drop=True) 
# drop=True --> avoid the old index being added as a column
df

Rank                                             Major Degree Type  \
0      -                                                 -           -   
1      1                             Petroleum Engineering   Bachelors   
2      2      Operations Research & Industrial Engineering   Bachelors   
3      3  Electrical Engineering & Computer Science (EECS)   Bachelors   
4      4                                Interaction Design   Bachelors   
..   ...                                               ...         ...   
823  823                                 Outdoor Education   Bachelors   
824  824                         Early Childhood Education   Bachelors   
825  825                                     Mental Health   Bachelors   
826  826                                 Medical Assisting   Bachelors   
827  827                                     Metalsmithing   Bachelors   

                                      Early Career Pay  \
0    Median salary for alumni with 0-5 years experi...   
1                                              $93,200   
2                                              $84,800   
3                                             $108,500   
4                                              $68,300   
..                                                 ...   
823                                            $37,400   
824                                            $36,100   
825                                            $36,900   
826                                            $36,000   
827                                            $40,000   

                                        Mid-Career Pay  \
0    Median salary for alumni with 10+ years experi...   
1                                             $187,300   
2                                             $170,400   
3                                             $159,300   
4                                             $155,800   
..                                                 ...   
823                                            $46,300   
824                                            $45,400   
825                                            $45,000   
826                                            $44,800   
827                                            $40,300   

                                        % High Meaning  
0    % alumni who say their work makes the world a ...  
1                                                  67%  
2                                                  28%  
3                                                  46%  
4                                                  55%  
..                                                 ...  
823                                                52%  
824                                                78%  
825                                                  -  
826                                                  -  
827                                                33%  

[828 rows x 6 columns]

---

## Initial remarks

1. There are 827 different Majors.

2.   Row 0 contains definitions for some of the column lables --> **Delete row 0.** ✅
    
3.   The index is repeating itself --> **Reset index.** ✅

4.   'Rank': The majors are ranked by "Mid-Career Pay"-values. Majors with the same Mid-Career Pay, have the same ranking. Do not need this column --> **Delete 'Rank'** ✅

5.   'Degree Type': There are only Bachelors. --> **Delete 'Degree Type'.** ✅

6.   There are some missing values, represented by "-".

7.   Check datatypes.

---

In [11]:
df.shape

(828, 6)

In [12]:
# (2) Delete row 0.
df = df.drop([0]) 
df.shape

(827, 6)

In [13]:
# (4)(5) Delete two columns.
df = df.drop(labels=['Rank', 'Degree Type'], axis=1) 
df

Major Early Career Pay  \
1                               Petroleum Engineering          $93,200   
2        Operations Research & Industrial Engineering          $84,800   
3    Electrical Engineering & Computer Science (EECS)         $108,500   
4                                  Interaction Design          $68,300   
5                                   Public Accounting          $59,800   
..                                                ...              ...   
823                                 Outdoor Education          $37,400   
824                         Early Childhood Education          $36,100   
825                                     Mental Health          $36,900   
826                                 Medical Assisting          $36,000   
827                                     Metalsmithing          $40,000   

    Mid-Career Pay % High Meaning  
1         $187,300            67%  
2         $170,400            28%  
3         $159,300            46%  
4         $155,800            55%  
5         $147,700            47%  
..             ...            ...  
823        $46,300            52%  
824        $45,400            78%  
825        $45,000              -  
826        $44,800              -  
827        $40,300            33%  

[827 rows x 4 columns]

---
**5. There are some missing values, represented by "-".**

✅ Replace '-' with '-1'.

**6. Check data-types**

✅ Make sure the numbers are int or float.

---

In [14]:
df.dtypes

Major               object
Early Career Pay    object
Mid-Career Pay      object
% High Meaning      object
dtype: object

In [18]:
num_df = df.replace('\$|,|%', '', regex=True)
num_df.head(5)

Major Early Career Pay  \
1                             Petroleum Engineering            93200   
2      Operations Research & Industrial Engineering            84800   
3  Electrical Engineering & Computer Science (EECS)           108500   
4                                Interaction Design            68300   
5                                 Public Accounting            59800   

  Mid-Career Pay % High Meaning  
1         187300             67  
2         170400             28  
3         159300             46  
4         155800             55  
5         147700             47

In [19]:
num_df['Early Career Pay'] = num_df['Early Career Pay'].astype(float)
num_df['Mid-Career Pay'] = num_df['Mid-Career Pay'].astype(float)
num_df['% High Meaning'] = num_df['% High Meaning'].replace(['-'], [-1]).astype(float)  
# Replace '-' with -1, and convert column to integers.

In [20]:
num_df.dtypes

Major                object
Early Career Pay    float64
Mid-Career Pay      float64
% High Meaning      float64
dtype: object

In [22]:
num_df.head(5)

Major  Early Career Pay  \
1                             Petroleum Engineering           93200.0   
2      Operations Research & Industrial Engineering           84800.0   
3  Electrical Engineering & Computer Science (EECS)          108500.0   
4                                Interaction Design           68300.0   
5                                 Public Accounting           59800.0   

   Mid-Career Pay  % High Meaning  
1        187300.0            67.0  
2        170400.0            28.0  
3        159300.0            46.0  
4        155800.0            55.0  
5        147700.0            47.0

---

# Styling

``df.style``


``df.style.format()``

keyword arguments:

*   precision: int= 0       
    * \# of decimals.
*   thousands: str= ' '
    * thousands separator, f.eks. ' ' (space).
*   formatter={
          ('column name'): 'formatting string'
}
    * Where formatting string is like an f-string, with python formatting-syntax. '$ {}'

In [23]:
df.describe()  # Without styling

Major Early Career Pay Mid-Career Pay % High Meaning
count                     827              827            827            827
unique                    827              328            506             72
top     Petroleum Engineering          $47,700        $85,900              -
freq                        1               12              7             64

In [24]:
df.describe().style.format(precision=0, thousands=' ')
# With no decimals and a space as thousands separator.

In [25]:
df = num_df
df.describe().style.format(precision=0, thousands=' ')
# With no decimals and space as thousands separator.

In [26]:
df.head(5).style.format(thousands=' ',
                        formatter={
                            ('Early Career Pay'): '$ {:,.0f}',
                            ('Mid-Career Pay'): '$ {:.0f}',
                            ('% High Meaning'): '{:.0f} %'
                        })
# If ',' in the formatting-string is included, 
# the thousands separator defined above is used.

In [27]:
# background_gradient requires matplotlib.

def make_pretty(styler):
    styler.set_caption("Highest Paying Jobs With a Bachelor’s Degree")
    styler.format(thousands=' ',
                        formatter={
                            ('Early Career Pay'): '$ {:,.0f}',
                            ('Mid-Career Pay'): '$ {:,.0f}',
                            ('% High Meaning'): '{:.0f} %'
                        })
    #styler.background_gradient(axis=None, vmin=147700, vmax=187300, cmap='YlGnBu')
    styler.background_gradient(axis=None, cmap='Blues')
    return styler

df.loc[1:5, 'Major':'Mid-Career Pay'].style.pipe(make_pretty)

---

# Statistics



## Early Career Pay

### **Majors with lowest Early Career Pay:**

In [ ]:
low_ecp = df.sort_values(by='Early Career Pay', ascending=True, ignore_index=True).head(5)
low_ecp_df = low_ecp.loc[0:,['Major', 'Early Career Pay']]
low_ecp_df.style.pipe(make_pretty)

### **Majors with highest Early Career Pay:**

In [ ]:
high_ecp = df.sort_values(by='Early Career Pay', ascending=False, ignore_index=True).head(5)
high_ecp.style.pipe(make_pretty)

In [ ]:
high_ecp_df = high_ecp.loc[0:, ['Major', 'Early Career Pay']]
high_ecp_df.style.pipe(make_pretty).hide()

In [ ]:
high_ecp_plot = high_ecp.head(5).plot(x='Major', y='Early Career Pay', kind='bar', 
                              color=['indianred', 'coral', 'gold', 'darkcyan', 'steelblue'])
high_ecp_plot

### **Highest and lowest Early career pay**

In [ ]:
high_low_ecp_df = pd.concat([low_ecp, high_ecp])
high_low_ecp_df = high_low_ecp_df.sort_values(by='Early Career Pay').reset_index()
high_low_ecp_table = high_low_ecp_df.loc[0:,['Major', 'Early Career Pay']]
high_low_ecp_table.style.pipe(make_pretty).hide()

In [ ]:
colors = ['#00429d', '#2855a6', '#3d68ae', '#4f7cb7', '#6090bf', '#70a4c7', '#81b9cf', '#94cdd6', '#abe2dc', '#d9f1de']
c = list(reversed(colors))
# Find nice colors at: https://gka.github.io/palettes
high_low_ecp_plot = high_low_ecp_df.plot(x='Major', y='Early Career Pay', kind='barh', color=c)

In [ ]:
ecp = high_low_ecp_df['Early Career Pay'].to_list()
mcp = high_low_ecp_df['Mid-Career Pay'].to_list()
index = high_low_ecp_df['Major']
ecp_df = pd.DataFrame({
    'Early career pay': ecp,
    'Mid-Career Pay': mcp
}, index=index)
# ecp_ax = ecp_df.plot.bar(rot=90, color=['#70a4c7', '#3d68ae'])

In [ ]:
ecp_ax2 = ecp_df.plot.barh(color=['#70a4c7', '#3d68ae'])

## Mid-Career Pay

In [ ]:
low_mcp_df = df.sort_values(by='Mid-Career Pay', ascending=True).head(5)
high_mcp_df = df.sort_values(by='Mid-Career Pay', ascending=False).head(5)

low_high_mcp_df = pd.concat([low_mcp_df, high_mcp_df])
low_high_mcp_df = low_high_mcp_df.sort_values(by='Mid-Career Pay')
low_high_mcp_df.style.pipe(make_pretty)

In [ ]:
mcp_df = pd.DataFrame({'Early career pay': low_high_mcp_df['Early Career Pay'].to_list(),
                       'Mid-Career Pay': low_high_mcp_df['Mid-Career Pay'].to_list()}, 
                      index=low_high_mcp_df['Major'])
mcp_ax = mcp_df.plot.barh(color=['#94cdd6', '#2855a6'])

# Highest increase in median salary from early to mid-career.

In [ ]:
# increase_col = df['Mid-Career Pay'] - df['Early Career Pay']
# df.insert(3, 'Increase', increase_col)

In [ ]:
high_increase_df = df.sort_values(by='Increase', ascending=False).head(10)
# pd.options.display.float_format = '{:,.2f}'.format  # Format output numbers.
high_increase_df

# Next
Work with bar(h) plots. E.g. add x-axis labels, units, header, figure text (?).
Make functions for styling different things.